cmake/checks/check_02_compiler_features.cmake

```diff
@@ -338,7 +338,7 @@ RESET_CMAKE_REQUIRED()
 
 IF(DEAL_II_COMPILER_HAS_CXX14_ATTRIBUTE_DEPRECATED)
   SET(DEAL_II_DEPRECATED "[[deprecated]]")
-ELSEIF(DEAL_II_COMPILER_HAS_ATTRIBUTE_DEPRECATED)
+ELSEIF(DEAL_II_COMPILER_HAS_ATTRIBUTE_DEPRECATED AND NOT DEAL_II_WITH_CUDA)
   SET(DEAL_II_DEPRECATED "__attribute__((deprecated))")
```

Comment:
This disables the attributes altogether if you use CUDA, whether or not a piece of code is compiled with gcc for the host or nvcc for the target. An alternative could have been to just 
```
#ifdef compile_for_host
  #define DEAL_II_DEPRECATED previous def
#else // compiling for the target with nvcc
  #define DEAL_II_DEPRECATED  // nothing
#endif
```

Reply:
Yes, I was aware of this. Clearly, the advantage is that all the other places where we use ` DEAL_II_DEPRECATED` still work. The disadvantage is that we might run into this problem repeatedly and bloat the code.
In case we have support for `[[deprecated]]`, the approach I decided to follow in this PR also marks this alias deprecated whereas this would not be the case with the preprocessor check.

Reply:
I meant to put the `#ifdef` into `config.h`. You won't have to annotate things for each function/variable/declaration in question, just one place.

Reply:
Ah, OK.